src/language/branches.md

```diff
@@ -0,0 +1,295 @@
+  ```dart
+  switch (number) {
+    case 1:           // Matches if number == 1
+      print("one");     
+  }
+  ```
+- **List**, **map**, or **record patterns**: the elements are destructured if matched.
+  ```dart
+  switch (obj) {
+    case [a, b]:                  // Matches if obj is a list with two elements
+      print("$a and $b match");    
+  }
+  ```  
+- **Variable patterns**: the matched value binds to a new variable.
+  ```dart
+  switch ((1, 2)) {
+    case (var a, var b): ...     // Matches if 1 and 2 can be assigned to a and b
+  }
+  ```  
+- **Logical** or **relational patterns**: the object matches if the comparison to a constant returns `true`.
+  ```dart
+  switch (code) {
+    case >= first && <= last: ...    
+    // Matches if code's value is between the constant values of first and last
+  }
+  ```  
+- **Null check** or **assert patterns**: the match is first on whether the object is not null, then on the value.
+  ```dart
+  switch (maybeString) {
+    case var s?:                 
+    // Matches if maybe is not null, and then against the inner pattern "var s"
+  }
+  ```  
+- **Object patterns**: the match exposes getters on the object type.
+  ```dart
+  switch (Shape shape) {
+    case Rect(width: var w, height: var h): ...
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a case pattern become local variables, only available in the
+scope of that clause. 
+
+### Switch statements
+
+Non-empty `case` clauses implicitly jump to the end of the switch after completion.
+Other valid ways to end a non-empty `case` clause are a [`continue`][break],
+[`throw`][], or [`return`][] statement.
+
+Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:
+
+```dart
+var command = 'OPEN';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+  case 'PENDING':
+    executePending();
+  case 'APPROVED':
+    executeApproved();
+  case 'DENIED':
+    executeDenied();
+  case 'OPEN':
+    executeOpen();
+  default:
+    executeUnknown();
+}
+```
+
+Empty cases fall through to the next case. For an empty case that does not fall
+through, use [`break`][break] for its body:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED': // Empty case falls through.
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+For non-sequential fall-through, you can use a [`continue` statement][break] and
+a label:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+    continue nowClosed;
+  // Continues executing at the nowClosed label.
+
+  nowClosed:
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+### Switch expressions
+
+_Switch expressions_ write a result value determined by a switch
+directly into a variable or return statement. 
```

Comment:
You can use a switch expression anywhere an expression can appear so "directly into a variable or return statement" is probably too specific. For example, this is fine:

```dart
print(switch (n) { 1 => 'one', 2 => 'two', _ => 'other'});
```

How about:

> A switch expression is similar to a switch statement, but can be used where an expression is expected and produces a value based on the expression body of the case that matched.